### DISCUSSION TITLE
Magnetic field from one and more point charge(s)

### DISCUSSION THREAD

**Discussion by Georgy1987** [magpylib/magpylib#599] [Q&A] [answered]
Hello All! Is there a way to use Magpylib to simulate magnetic field from one and more point charges(s) [provided in A-m or Wb units] and "measure" it in tesla (or gauss) from known 3D coordinates of the observer position to known 3D coordinates of the charge position(s)?

If this is possible, what would be the most basic example of the code for such simulation?

Thanks lots!

**Comment by OrtnerMichael** [comment]
hey @Georgy1987 

sorry for the late reply. 

Yes its possible - but you have to clarify what you mean by point charge. The elementary magnetic charge is a dipole. The dipole moment is provided e.g. in A-m^2, not A-m! In Magpylib we use mT mm^3. Wb is a unit that is used to measure magnetic flux, i.e. 1Wb = 1 Tm^2 !

A simulation example is simple:

```python
import magpylib as magpy

obs_pos = (1,2,3)

# creates a dipole moment that points in z-direction with moment of 1000 in units of mT mm^3
dipole = magpy.misc.Dipole(moment = (0,0,1000))

# computes the magnetic field in units of mT at observer position
B = dipole.getB(obs_pos)

print(B)
```

does this help ?

**Reply by Georgy1987** [reply]
Hello, @OrtnerMichael . Forgive me for not being clear. I tried the code you suggested some time ago and it works great for its intended purpose (Thank you!). I was wondering if there is a way in Magpylib to simulate, for example, "a dipole" using two opposing poles with given pole strength (A-m or Wb) and their known coordinates in space, and measure the **resulting** magnetic field (H or B)  due to this "dipole" from known coordinates of the observer? So the input would be: + & - pole strength (two poles: positive and negative), coordinates of the poles, coordinates of the observer.

**Comment by OrtnerMichael** [comment, marked as answer]
hi @Georgy1987 
Well yeah sure thats possible, but you would have to provide the monopole field of each charge by yourself, which is not difficult. There is even an example in the docu that does exactly what our want, except with a quadrupole.

[Quadrupol Example](https://magpylib.readthedocs.io/en/latest/examples/examples_04_custom_source.html)

The only thing you would have to provide is the unit conversion.

**Reply by Georgy1987** [reply]
Thanks very much @OrtnerMichael ,

I was able to assemble the geometry I was looking for using the monopoles.
Can you please let me know the unit for the charge in that quadrupole code example? Is it kA-mm?

Thanks again, the visualization looks absolutely beautiful!

**Reply by OrtnerMichael** [reply]
hi @Georgy1987 

thanks :)

So the custom class works like this: You give a B-function (in this case the monopole field) and the class.getB will compute B according to your function. In the example the input is `charge` and `position`. The output is `charge*(position.T/dist**3).T`. The output units depend on the input units. To get the correct field out in terms of units for an electric point charge you need to include the electric field constant as follows:

```
# field of electric point charge
def monopole_field(charge, position):
    """ field point charge"""
    eps0 = 8.8541878128e-12
    position = np.array(position)
    dist = np.linalg.norm(position, axis=1)
    return charge/(4*np.pi*eps0)*(position.T/dist**3).T
```

Replace the function in the example with this one, then if your input is in SI units (position in meter, charge in Coulomb) your output will be in SI units as well, i.e. Volt / meter.

sincerely,
Michael
